src/Confluent.Kafka/Consumer.cs

```diff
@@ -177,37 +179,85 @@ private void LogCallback(IntPtr rk, SyslogLevel level, string fac, string buf)
                     throw new Exception("Unexpected rebalance callback on disposed kafkaHandle");
                 }
 
+                if (kafkaHandle.RebalanceProtocol == "COOPERATIVE" &&
+                    this.revokedOrLostHandlerIsFunc)
+                {
+                    throw new InvalidOperationException("Neither revoked nor lost partition handlers may return an updated assignment when a COOPERATIVE assignor is in use");
+                }
+
+                var partitions = SafeKafkaHandle.GetTopicPartitionOffsetErrorList(partitionsPtr).Select(p => p.TopicPartition).ToList();
+
                 if (err == ErrorCode.Local_AssignPartitions)
                 {
                     if (partitionsAssignedHandler == null)
                     {
-                        Assign(partitionAssignment.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                        if (kafkaHandle.RebalanceProtocol == "COOPERATIVE")
+                        {
+                            IncrementalAssign(partitions.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                        }
+                        else
+                        {
+                            Assign(partitions.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                        }
                         return;
                     }
 
                     lock (assignCallCountLockObj) { assignCallCount = 0; }
-                    var assignTo = partitionsAssignedHandler(partitionAssignment);
+                    var assignTo = partitionsAssignedHandler(partitions);
                     lock (assignCallCountLockObj)
                     {
                         if (assignCallCount > 0)
                         {
-                            throw new InvalidOperationException("Assign/Unassign must not be called in the partitions assigned handler.");
+                            throw new InvalidOperationException("(Incremental)Assign/Unassign must not be called in the partitions assigned handler");
                         }
                     }
-                    Assign(assignTo);
+
+                    if (kafkaHandle.RebalanceProtocol == "COOPERATIVE")
+                    {
+                        if (assignTo.Count() != partitions.Count())
+                        {
+                            throw new InvalidOperationException("The partitions assigned handler must not return a different set of topic partitions than it was provided");
+                        }
+
+                        int i = 0;
+                        foreach (var p in assignTo)
+                        {
+                            if (p.TopicPartition != partitions[i++])
```

Comment:
Are these two always sorted? If not, sort them.

Reply:
they are, provided the user doesn't change the ordering of what they are given if they choose to return a collection with different offsets. which usually they wouldn't. Ok, i'll sort the user returned collection. I guess only O(n) in the usual case, so no big deal.